packages/graph-explorer/src/core/StateProvider/localDb.ts

```diff
@@ -0,0 +1,223 @@
+import { z } from "zod";
+import { logger } from "../../utils";
+import { deserializeData, serializeData } from "./serializeData";
+import { fromFileToJson, toJsonFileData } from "../../utils/fileData";
+import saveAs from "file-saver";
+
+/*
+
+# Dev Notes
+
+The backup & restore process is a bit complex. Here is how the process breaks
+down.
+
+## Backup Process
+
+This process will create a JSON file with all the contents of the localForage
+storage, along with some extra metadata.
+
+1. Clean up any left over restored entries from previous failed restores.
+2. Gather all data from localForage storage.
+3. Wrap in backup envelope with some metadata about the backup.
+4. Transform data to a format that is safe to serialize to a file.
+5. Transform data in to binary blob data.
+6. Save to file
+
+## File Load Process
+
+This process will convert the binary data in the file to a JSON structure that
+can be restored. The actual restoration process is separate. This step is used
+as a validation of the backup data in the file.
+
+1. Transform binary blob data to file safe JSON data.
+2. Transform file safe JSON data to normal JSON data.
+3. Parse JSON data as backup data.
+
+## Restore Process
+
+Since we don't have transactions, we take a cautious approach to the restore
+process. We attempt to write all entries from the backup file to the localForage
+storage. If this succeeds, then we remove the old entries and rename the
+restored entries.
+
+If there is a failure during the writing of the restore entries then the
+existing database is left in a working state.
+
+1. Clean up any left over restored entries from previous failed restores.
+2. Write all the backup data entries to localForage with "restored" prefix.
+3. Delete all non-restored entries.
+4. Rename all restored entries by removing prefix.
+
+*/
+
+/** Backs up the localDb to a file */
+export async function saveLocalForageToFile(localDb: LocalDb) {
+  const backup = await createBackupData(localDb);
+  const serializedExportData = serializeData(backup);
+  logger.debug("Backup serialized for file", serializedExportData);
+  const fileData = toJsonFileData(serializedExportData);
+  logger.debug(`Converted data to JSON file blob`);
+  const fileName = "graph-explorer-config.json";
```

Comment:
nit:? Constants - upper case? e.g. `fileName` --> `BACKUP_FILE_NAME` ? Same for above.

Reply:
That's a good idea.

Reply:
After looking a bit closer at this one, I decided to keep the string where it is. This is the only place it is used and there's no risk if it changes. So the indirection of aa constant in another file, or even somewhere else in this file, is not useful.